### ISSUE TITLE
INTERNAL ERROR only onBecomeUnobserved shouldn't be called twice in a row

### ISSUE THREAD

**Issue by jamiewinder** [mobxjs/mobx#874]
I'm writing a utility for MobX which is meant to act as a lazily-evaluated value which expires (and is re-evaluated) over time. This is using an `Atom` behind the scenes so I can cancel my expiry / re-evaluation if it becomes un-observed. 

However, when my atom becomes unobserved, is seems to throw the error in the title which suggests its an internal MobX problem? Here is my code:

```typescript
export function lazyExpiringObservable<TExp, TValue>(
    expression: () => TExp,
    evaluate: (arg: TExp) => TValue,
    expiry: number
) {
    let currentValue: TValue;

    const atom = new Atom('lazyExpiringObservable', startTracking, stopTracking);

    // Whenever a value is set, an expiry timeout [re]starts which will reevaluate
    // the expression. This timeout is cancelled if the atom becomes unobserved
    let expiryTimeout = 0;
    function resetExpiryTimeout() {
        clearExpiryTimeout();
        expiryTimeout = setTimeout(() => {
            const nextValue = evaluate(expression());
            updateValue(nextValue);
            resetExpiryTimeout();
        }, expiry);
    }
    function clearExpiryTimeout() {
        if (expiryTimeout) {
            clearTimeout(expiryTimeout);
            expiryTimeout = 0;
        }
    }

    function updateValue(value: TValue) {
        currentValue = value;
        atom.reportChanged();
        resetExpiryTimeout();
    }

    // The expression is tracked, so if any observable it references changes,
    // the value will be re-evaluated and updated.
    let currentReaction: Lambda;
    function startTracking() {
        const value = evaluate(expression());
        updateValue(value);
        currentReaction = reaction(
            expression,
            (arg) => {
                const nextValue = evaluate(arg);
                updateValue(nextValue);
            }
        );
    }
    function stopTracking() {
        currentReaction();
        clearExpiryTimeout();
    }
    return {
        get(): TValue {
            if (atom.reportObserved()) {
                return currentValue;
            } else {
                throw new Error('lazyExpiringObservable accessed outside reaction');
            }
        }
    };
}
```

In practice, I have an `observer`-based React component which references the returned `get()` and will happily update if either the expression changes or the expiry timeout occurs. However, if I unmount the component I get the title error which is here in full:

```bash
Uncaught Error: [mobx] Invariant failed: INTERNAL ERROR only onBecomeUnobserved shouldn't be called twice in a row
    at invariant (index.js:8290)
    at ComputedValue.onBecomeUnobserved (index.js:6392)
    at endBatch (index.js:6762)
    at Reaction.dispose (index.js:6927)
    at Atom.stopTracking [as onBecomeUnobservedHandler] (index.js:73135)
    at Atom.onBecomeUnobserved (index.js:6358)
    at endBatch (index.js:6762)
    at Reaction.dispose (index.js:6927)
    at Atom.stopTracking [as onBecomeUnobservedHandler] (index.js:73135)
    at Atom.onBecomeUnobserved (index.js:6358)
```
This is logged preceded by a React 'warning' which may be related (likely an effect rather than a cause...?):
```bash
Warning: There is an internal error in the React performance measurement code. We did not expect componentWillUnmount timer to stop while no timer is still in progress for another instance. Please report this as a bug in React.
```
I'm trying to produce a small reproduction, but if anyone can spot the problem (or thinks it's an MobX issue) please let me know.

**Comment by jamiewinder** [comment]
I've still not reproduced it; having trouble replicating the complex model structure at work in my application. However, I think it's something to do with the disposal of the `reaction` in my `stopTracking` function; this is called by the `onBecomeUnobserved` event of the Atom. If I wrap this in a `setTimeout(() => currentReaction, 1)` it seems to fix it... is there something about an `Atom` becoming unobservable and disposing a reaction (in the same cycle) that might be incompatible?

**Comment by mweststrate** [comment]
@jamiewinder yeah that might very well be the case. For me to be able to understand the problem, could  you describe what the `lazyExpiringObservable` is intended to do with some example :)?

**Comment by jamiewinder** [comment]
@mweststrate 

Thanks! Essentially, it's trying to be `lazyObservable`, but with a time-based expiration of the data and a reaction around the parameters. I'm using it for a parameterized server request for some data, something like:

`/api/getcount?groups=xxx,yyy`

For which I wanted the following behaviours:

- Won't be resolved until it's observed
- Whenever evaluated, start a timer to 'expire' (re-request) the data after the configured timeout
- If the parameters ('groups' in this case) change, re-evaluate immediately
- When unobserved, stop any pending timeouts

So for the API I described, you could do:

```typescript
const serverCount = lazyExpiringObservable(
    // the observed expression, normally involving some observables. if these change, the value is reevaluated immediately
    () => appState.groupNames,
    // the evaluation of the expression (similar to the relationship between expression and effect in 'reaction'
    (groups) => fromPromise('/api/getcount?groups=' + groups.join(',')),
    // automatically expire and reevaluate after this time once a value is set, unless this observable isn't
    // observed
    60000
);
```
Does that make sense? I don't think anything like this exists in the utils library.

I'm still having trouble making a reproduction, but annoyingly it seems to work in all my attempts so far. :( 

I've a feeling something somewhere is getting in a loop, causing the React performance measurement code to abort (which is happening during the un-mounting of the `observer` component).

Thanks again.

**Comment by ronag** [comment]
We have encountered the same error after upgrading to 3.0. With the following pattern.

'''ja
computed(() => autorun(() => foo(atom.get()))
'''

**Comment by ronag** [comment]
Possibly related #898

**Comment by mweststrate** [comment]
@ronag / @jamiewinder could you check whether this build fixes the issue? mobx@3.1.8-fix874

**Comment by mweststrate** [comment]
Released 3.1.9 which should mitigate this issue. If it doesn't, feel free to reopen
